Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1068 @@
+    }
+}
+
+// Steals a reference to obj
+static void
+sym_set_type_from_const(_Py_UOpsSymType *sym, PyObject *obj)
+{
+    PyTypeObject *tp = Py_TYPE(obj);
+
+    if (tp == &PyLong_Type) {
+        sym_set_type(sym, PYLONG_TYPE, 0);
+    }
+    else if (tp == &PyFloat_Type) {
+        sym_set_type(sym, PYFLOAT_TYPE, 0);
+    }
+    else if (tp == &PyUnicode_Type) {
+        sym_set_type(sym, PYUNICODE_TYPE, 0);
+    }
+
+    if (tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
+        PyDictOrValues *dorv = _PyObject_DictOrValuesPointer(obj);
+
+        if (_PyDictOrValues_IsValues(*dorv) ||
+            _PyObject_MakeInstanceAttributesFromDict(obj, dorv)) {
+            sym_set_type(sym, GUARD_DORV_VALUES_INST_ATTR_FROM_DICT_TYPE, 0);
+
+            PyTypeObject *owner_cls = tp;
+            PyHeapTypeObject *owner_heap_type = (PyHeapTypeObject *)owner_cls;
+            sym_set_type(
+                sym,
+                GUARD_KEYS_VERSION_TYPE,
+                owner_heap_type->ht_cached_keys->dk_version
+            );
+        }
+
+        if (!_PyDictOrValues_IsValues(*dorv)) {
+            sym_set_type(sym, GUARD_DORV_VALUES_TYPE, 0);
+        }
+    }
+}
+
+
+static inline _Py_UOpsSymType*
+sym_init_unknown(_Py_UOpsAbstractInterpContext *ctx)
+{
+    return _Py_UOpsSymType_New(ctx,NULL);
+}
+
+// Takes a borrowed reference to const_val.
+static inline _Py_UOpsSymType*
+sym_init_const(_Py_UOpsAbstractInterpContext *ctx, PyObject *const_val, int const_idx)
+{
+    assert(const_val != NULL);
+    _Py_UOpsSymType *temp = _Py_UOpsSymType_New(
+        ctx,
+        const_val
+    );
+    if (temp == NULL) {
+        return NULL;
+    }
+    sym_set_type(temp, TRUE_CONST, 0);
+    return temp;
+}
+
+static _Py_UOpsSymType*
+sym_init_push_null(_Py_UOpsAbstractInterpContext *ctx)
+{
+    if (ctx->frequent_syms.push_nulL_sym != NULL) {
+        return ctx->frequent_syms.push_nulL_sym;
+    }
+    _Py_UOpsSymType *null_sym = sym_init_unknown(ctx);
+    if (null_sym == NULL) {
+        return NULL;
+    }
+    sym_set_type(null_sym, NULL_TYPE, 0);
+    ctx->frequent_syms.push_nulL_sym = null_sym;
+    return null_sym;
+}
+
+static inline bool
+sym_is_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    if ((sym->types & (1 << typ)) == 0) {
+        return false;
+    }
+    return true;
+}
+
+static inline bool
+sym_matches_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ, uint64_t refinement)
+{
+    if (!sym_is_type(sym, typ)) {
+        return false;
+    }
+    if (typ <= MAX_TYPE_WITH_REFINEMENT) {
+        return sym->refinement[typ] == refinement;
+    }
+    return true;
+}
+
+static uint64_t
+sym_type_get_refinement(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    assert(sym_is_type(sym, typ));
+    assert(typ <= MAX_TYPE_WITH_REFINEMENT);
+    return sym->refinement[typ];
+}
+
+
+static inline bool
+op_is_end(uint32_t opcode)
+{
+    return opcode == _EXIT_TRACE || opcode == _JUMP_TO_TOP;
```

Comment:
I have a suspicion that this should also test for `_JUMP_ABSOLUTE`. And possibly one or two other places.

Reply:
`_JUMP_ABSOLUTE` never appears in the trace naturally except after written by us.

Reply:
Oh, subtle. (Maybe *too* subtle.)

I do think there may be another trap: `remove_unneeded_uops()` breaks out of its loop when it finds `_JUMP_TO_TOP` or `_EXIT_TRACE`. It looks like that *will* encounter `_JUMP_ABSOLUTE` and fail to exit there (maybe that's the reason why you have to fill the rest of the buffer with `_NOP`).

Reply:
Oh you are spot on! Nice. Thanks for catching that.